### PR TITLE
Improving support for `jax-jit` interface with `default.mixed`

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** Adds more tests for ensuring  `default.mixed` works well with the `jax-jit` interface 

**Description of the Change:** Added support for computing samples in the `jax-jit` case and more generalized tests. 

**Benefits:** 

**Possible Drawbacks:** N/A

**Related GitHub Issues:** N/A

### CODE REVIEW

**Comment by obliviateandsurrender** [comment]
[sc-60973]

**Review by trbromley — COMMENTED** [review]
Thanks @obliviateandsurrender! The Shortcut story mentions also making sure we have solid tests for all of the interfaces. Do you think this is already satisfied, or are you planning to make a follow up PR?

File: pennylane/_qubit_device.py [hunk 1]
```diff
@@ -867,15 +867,28 @@ def sample_basis_states(self, number_of_states, state_probability):
             )
 
         shots = self.shots
-
         basis_states = np.arange(number_of_states)
+        # pylint:disable = import-outside-toplevel
+        if qml.math.is_abstract(state_probability):
+            import jax
+
+            jax.config.update("jax_enable_x64", True)
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/_qubit_device.py:875]: Included it to silence warnings related to it not being enabled. It could be skipped.

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -223,6 +223,10 @@
 * `qml.transforms.hamiltonian_expand` can now handle multi-term observables with a constant offset.
   [(#5414)](https://github.com/PennyLaneAI/pennylane/pull/5414)
 
+* `default.mixed` has improved support for sampling-based measurements with non-numpy interfaces.
+  [(#5514)](https://github.com/PennyLaneAI/pennylane/pull/5514)
+  [(#5530)](https://github.com/PennyLaneAI/pennylane/pull/5530)
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply doc/releases/changelog-dev.md:228]: Changes are mainly related to sampling, so I think yes.

**Comment by obliviateandsurrender** [comment]
> The Shortcut story mentions also making sure we have solid tests for all of the interfaces

As far as I could see, I think this is already satisfied. I would still go over them once more.

**Review by Jaybsoni — COMMENTED** [review]
Looks great, I would just like to see more integration tests with Qnode. Some of the tests you added act on the qscript level, but it would be nice to make sure nothing breaks when executing through the qnode.

File: pennylane/_qubit_device.py [hunk 3]
```diff
@@ -867,15 +867,28 @@ def sample_basis_states(self, number_of_states, state_probability):
...
+            key = jax.random.PRNGKey(np.random.randint(0, 2**31))
+            if jax.numpy.ndim(state_probability) == 2:
+                return jax.numpy.array(
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/_qubit_device.py:878]: Yep.

File: tests/devices/test_default_mixed_jax.py [hunk 4]
```diff
@@ -774,3 +838,45 @@ def circuit(p):
             expected.append(circuit(x_indiv))
 
         assert np.allclose(expected, res)
+
+    @pytest.mark.parametrize("gradient_func", [qml.gradients.param_shift, "device", None])
+    def test_tapes(self, gradient_func):
+        """Test that jitted execution works with tapes."""
```
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply tests/devices/test_default_mixed_jax.py:844]: Both for coverage and completeness. I went through the existing `jax-jit` tests in the suite and added tests based on them.

[the remark this answers is not in the document]
Comment (Jaybsoni) [reply tests/devices/test_default_mixed_jax.py:844]: Awesome 👍🏼

File: tests/devices/test_default_mixed_jax.py [hunk 5]
```diff
@@ -113,6 +113,70 @@ def circuit(state_ini):
...
+
+        dev = qml.device("default.mixed", wires=1, shots=10)
+
+        @jax.jit
+        def wrapper(x):
+            with qml.queuing.AnnotatedQueue() as q:
+                qml.RX(x, wires=0)
+                qml.expval(qml.PauliZ(0))
+            tape = qml.tape.QuantumScript.from_queue(q)
+            return qml.execute([tape], dev, gradient_fn=gradient_func)
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply tests/devices/test_default_mixed_jax.py:159]: Yes please,

File: pennylane/_qubit_device.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/_qubit_device.py:881]: Any thoughts @mlxd?

**Review by mudit2812 — COMMENTED** [review]
One open question so I'm holding off on approval for now, but looks good!

File: pennylane/_qubit_device.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (obliviateandsurrender) [reply pennylane/_qubit_device.py]: I have removed this. 👍

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -282,10 +282,12 @@
 
 * `default.mixed` has improved support for sampling-based measurements with non-numpy interfaces.
   [(#5514)](https://github.com/PennyLaneAI/pennylane/pull/5514)
+  [(#5530)](https://github.com/PennyLaneAI/pennylane/pull/5530)
 
 * The `qml.qchem.hf_state` function is upgraded to be compatible with the parity and Bravyi-Kitaev bases.
   [(#5472)](https://github.com/PennyLaneAI/pennylane/pull/5472)
 
+
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:290]:
```suggestion
```

**Review by mudit2812 — APPROVED** [review]

**Review by Jaybsoni — APPROVED** [review]
💯

**Comment by astralcai** [comment]
@obliviateandsurrender Hi, we have a feature freeze for the v0.36 release starting next Monday. When you feel that this PR is ready to be merged, please label it merge-ready. I will be manually managing the merge queue over the next few days to make sure everything gets merged in time for feature freeze without wasting too much time on repeated CI runs.
